pypeit/coadd3d.py

```diff
@@ -0,0 +1,1772 @@
+
+        """
+        if primary_hdr is None:
+            primary_hdr = io.initialize_header()
+        # Build the header
+        if self.head0 is not None and self.PYP_SPEC is not None:
+            spectrograph = load_spectrograph(self.PYP_SPEC)
+            subheader = spectrograph.subheader_for_spec(self.head0, self.head0)
+        else:
+            subheader = {}
+        # Add em in
+        for key in subheader:
+            primary_hdr[key] = subheader[key]
+        # Do it
+        super(DataCube, self).to_file(ofile, primary_hdr=primary_hdr, hdr=hdr, **kwargs)
+
+    @classmethod
+    def from_file(cls, ifile):
+        """
+        Over-load :func:`~pypeit.datamodel.DataContainer.from_file`
+        to deal with the header
+
+        Args:
+            ifile (str):  Filename holding the object
+        """
+        with io.fits_open(ifile) as hdu:
+            # Read using the base class
+            self = super().from_hdu(hdu)
+            # Internals
+            self.filename = ifile
+            self.head0 = hdu[1].header  # Actually use the first extension here, since it contains the WCS
+            # Meta
+            self.spectrograph = load_spectrograph(self.PYP_SPEC)
+            self.spect_meta = self.spectrograph.parse_spec_header(hdu[0].header)
+        return self
+
+    @property
+    def ivar(self):
+        """
+        Utility function to compute the inverse variance cube
+        """
+        return utils.inverse(self.sig**2)
+
+    @property
+    def wcs(self):
+        """
+        Utility function to provide the world coordinate system of the datacube
+        """
+        return wcs.WCS(self.head0)
+
+
+class CoAdd3D:
+    """
+    Main routine to convert processed PypeIt spec2d frames into
+    DataCube (spec3d) files. This routine is only used for IFU
+    data reduction.
+
+    Algorithm steps are detailed in the coadd routine.
+    """
+    # Superclass factory method generates the subclass instance
+    @classmethod
+    def get_instance(cls, spec2dfiles, opts, spectrograph=None, par=None, det=1, overwrite=False,
+                     show=False, debug=False):
+        """
+        Instantiate the subclass appropriate for the provided spectrograph.
+
+        The class to instantiate must match the ``pypeline``
+        attribute of the provided ``spectrograph``, and must be a
+        subclass of :class:`CoAdd3D`; see the parent class
+        instantiation for parameter descriptions.
+
+        Returns:
+            :class:`CoAdd3D`: One of the subclasses with
+            :class:`CoAdd3D` as its base.
+        """
+
+        return next(c for c in cls.__subclasses__()
+                    if c.__name__ == (spectrograph.pypeline + 'CoAdd3D'))(
+                        spec2dfiles, opts, spectrograph=spectrograph, par=par, det=det, overwrite=overwrite,
+                        show=show, debug=debug)
+
+    def __init__(self, files, opts, spectrograph=None, par=None, det=None, overwrite=False,
```

Comment:
Can you change this name to spec2dfiles for consistency with naming above and also for clarity.

Reply:
Ah, yes, I meant to do this - good catch.